content/blog/post-9-agentic-ai-with-agno-ollama-and-local-llama3.md

```diff
@@ -0,0 +1,255 @@
+    <tr>
+      <td><code>model=Ollama(...)</code></td>
+      <td>Connects to a local LLaMA model via Ollama</td>
+    </tr>
+    <tr>
+      <td><code>description</code></td>
+      <td>Agent personality and behavior</td>
+    </tr>
+    <tr>
+      <td><code>markdown=True</code></td>
+      <td>Outputs markdown-formatted content</td>
+    </tr>
+  </tbody>
+</table>
+
+## No Framework? No Problem.
+
+Frameworks like AGNO offer orchestration, but what if you're running in:
+
+* Air-gapped networks
+* Lightweight environments
+* Custom experimental setups?
+
+Here’s how to build a raw agent pipeline using just:
+
+* Ollama for LLM
+* DuckDuckGo search for tool use
+* Custom prompt logic
+
+## Full python Agent pipeline
+
+```python
+from ollama import Client
+
+from duckduckgo_search import DDGS
+
+ollama_client = Client(host='http://localhost:11434')
+
+### Tool: Web search via DuckDuckGo
+
+def duckduckgo_search(query, max_results=5):
+
+with DDGS() as ddgs:
+
+     return \[r for r in ddgs.text(query, region="wt-wt", safesearch="off", max_results=max_results)]
+```
+
+### Pipeline logic
+
+```python
+def agent_pipeline(user_input):
+
+if "what" in user_input.lower() or "happening" in user_input.lower():
+
+     print("\[Tool] Searching DuckDuckGo...")
+
+     search_results = duckduckgo_search(user_input)
+
+     summary = "\n".join(\[f"- {r['title']}: {r\['href']}" for r in search_results])
+
+else:
+
+     summary = ""
+
+if summary:
+
+     system_prompt = (
+
+         "You are an assistant. I found these recent search results:\n"
+
+         f"{summary}\n"
+
+         "Now generate a helpful answer for the user question below.\n"
+
+         f"User question: {user_input}"
+
+     )
+
+else:
+
+     system_prompt = f"You are a helpful assistant. Answer this: {user_input}"
+
+response = ollama_client.chat(
+
+     model='llama3.2',
+
+     messages=\[{"role": "user", "content": system_prompt}]
+
+)
+
+print("\n\[Agent Response]:")
+
+print(response\['message']\['content'])
+```
+
+### Example outputs
+
+```python
+agent_pipeline("What's happening in New York?")
+
+agent_pipeline("Tell me a joke.")
+```
+
+## Breakdown of components
+
+<table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%;">
+  <thead style="background-color:#f2f2f2">
+    <tr>
+      <th>Component</th>
+      <th>Role</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><code>LLM Backend</code></td>
+      <td>Ollama running LLaMA 3.2 locally</td>
+    </tr>
+    <tr>
+      <td><code>Tool</code></td>
+      <td>DuckDuckGo search for real-time info</td>
+    </tr>
+    <tr>
+      <td><code>Orchestration</code></td>
+      <td>Custom Python logic</td>
+    </tr>
+    <tr>
+      <td><code>Agentic Behavior</code></td>
+      <td>Manual Think → Act → Reflect implementation</td>
+    </tr>
+  </tbody>
+</table>
+
+This pattern gives you full control with zero cloud dependency — and forms the base for private AI workflows.
```

Comment:
Do you mean that it "forms the base for" or "forms the basis for"? I think it's the latter.